### ISSUE TITLE
Question: about Take / Threading

### ISSUE THREAD

**Issue by JanEggers** [dotnet/reactive#1975]
im trying to build a scheduler that queues work based on the item published.

the basics of that scheduler seem to work but when I do a take 100000 the resulting list most of the time contains less elements in it and im not sure why thats the case. I also tried using your ObserveOn Extension but I was not able to figure out how to implement The IScheduler as it does not provide the concrete values passed into the source observable

> Which library version?
6.0

> Do you have a code snippet or project that reproduces the problem?

https://github.com/JanEggers/Playground/blob/master/Playground.Core.Test/DispatchByTest.cs

so any hints whats wrong are welcome if you think its out of scope just close the issue.

**Comment by idg10** [comment]
I see your `FancyScheduler<TSource, TKey>` type implements `IObservable<T>`, not `IScheduler`, so it isn't a "scheduler" in the sense that Rx defines that word.

It's not entirely clear to me what you're trying to achieve here. Looking at the code, I _think_ what you're doing is partitioning the input based on the `keyselector` callback (defaulting to one partition for each distinct source value) and for each partition, you're spinning up a long-running task (which, in practice, will start a dedicated thread) that is responsible for invoking `OnNext` on observers for all the source elements in that particular partition.

E.g., suppose a source only ever produces the values 1 and 2, and we use the overload with the default logic, so each gets its own partition. The effect will be that you have two long-running tasks (and in practice, that will mean 2 threads), one of which performs all of the `OnNext(1)` calls on subscribers, and the other of which performs all the `OnNext(2)` calls on subscribers.

It looks like the design is intended to allow these to run in parallel. When you receive an incoming notification, you put it into a `Channel` specific to its partition. Each of the long-running tasks is pulling messages out of exactly one channel.

So if I understand the intent correctly, I think what you're trying to do is enable your operator's `OnNext` to complete almost immediately (it just drops the notification into the appropriate channel and then returns). If any of the the downstream observers have a slow `OnNext`, that will block only that partition, enabling other partitions to continue to deliver items.

Is my understanding correct?

If so, this is definitely illegal in Rx.

Rx defines some fundamental rules for code calling into an `IObserver<T>` implementation. These are described in section 4 of https://github.com/dotnet/reactive/blob/main/Rx.NET/Documentation/Rx%20Design%20Guidelines.pdf

The wording is a little odd: the sections state the assumptions that the Rx library makes. Section 4.2 states that the Rx library assumes that "observer instances are called in a serialized fashion". What that means is that for an single `IObserver<T>` instance, you must not have more than one call into it in progress at once.

For example, suppose you have called `OnNext` on one of the observers that have subscribed to your `FancyScheduler<TSource, TKey>`, and suppose that call has not yet returned. And now suppose you decide you want to make another call into it. The Rx library assumes that calls occur "in a serialized fashion", i.e., one after another. So **you are not allowed** to make another call to `OnNext` until the one that was already in progress returns. (Also, you're not allowed to call either `OnError` or `OnCompleted` if there is a call in progress to `OnNext` that has not yet returned.)

If you do not conform to these rules, the behaviour of the operators in the Rx library is undefined.

> when I do a take 100000 the resulting list most of the time contains less elements in it and im not sure why thats the case

The short answer is: you broke the rules, so the behaviour is undefined. If you make an `OnNext` call to the `IObserver<T>` that `Take` passed when subscribing to your source, and you don't wait for that call to return before making another call to `OnNext`, you are violating an assumption that `Take` makes. So it's allowed to do literally anything at this point.

So be a little more specific, because the rules enable the `Take` operator to presume that there will be no concurrent calls to any single instance of the `IObserver<T>` it supplies, the `Take` operator does not make any attempt at synchronization. It does not use synchronization primitives or `Interlocked`. It just uses the ordinary C# decrement syntax to count how many items it has received.

This is the correct implementation strategy because it never subscribes to multiple sources, and the rules of Rx enable it to assume that a single source will not ask it to handle multiple concurrent calls. Therefore, any code designed to cope with concurrent calls would be a waste of CPU cycles.

To be even more specific, I have two hypotheses about why you might be seeing this specific behaviour:

1. concurrent execution of a non-thread-safe decrement inside the `Take` operator is causing the counting to go wrong
2. you've not done anything to ensure that you drain the channels before stopping

I think 2 is more likely to be correct, because it seems more likely to me that 1 would cause `Take` to accept too many items, not too few. My thinking with 2 is that your `OnCompleted` implementation immediately invokes `OnCompleted` on all observers without doing anything to ensure that any of the long-running tasks have come to a halt. If you are pushing elements through the system as fast as you possible can, it's very likely that there will be unprocessed elements in the channels. Your channel handlers may well eventually process them, but they will now be breaking a different fundamental Rx rule: once you've called `OnCompleted` you're not allowed to call any other methods on that `IObserver<T>`.

The Rx library's implementations of `IObserver<T>` are defensively coded: if you do break this rule, they just ignore you. So if your code calls `OnNext` on an observer after having already called `OnCompleted` (which seems very likely, given this code), Rx just ignores all those calls. It is entitled to do this because you have broken the rules by making those calls, so the behaviour is undefined.


> I was not able to figure out how to implement The IScheduler as it does not provide the concrete values passed into the source observable

Yes, this is true. `IScheduler` is, by design, entirely unaware of the nature of the work is is being asked to schedule. But even if this weren't the case, it wouldn't help you here because the thing you are trying to achieve violates one of the most fundamental rules in Rx.

I think you need a fundamental change of approach here.

Since I'm not entirely sure why you want your code to do this, the following might not work. But my assumption is you have a scenario where the amount of processing work done for each item is non-trivial, and you want to enable multithreaded processing of such items.

The first thing I would ask is whether trying to achieve this inside of Rx is even the right solution. Very often when people attempt to do this sort of thing, it's because the final step of processing is slow. So they have an Rx query, and then they subscribe to that to receive the notifications that emerge from Rx, and then they want to do some quite slow work for each notification. Here's a simplified illustration of that sort of thing:

```cs
IObservable<Data> dataItems = myService.GetDataStream();

dataItems.Subscribe(
   data =>
   {
      myDatabase.Store(data);
   });
```

Here we're writing each message into some data store. That might be slow. Perhaps your data store offers higher throughput, but you'd need to have multiple connections open to it and have multiple concurrent writes in progress to exploit that throughput.

This is the point at which people often reach for multithreading, and they may ask "So how can I make Rx invoke that callback concurrently."

But that's the wrong approach because it will necessarily entail breaking the rules of Rx.

For that simple example above, your best bet is often to use TPL Dataflow, which has integrated support for Rx. If you define an `ActionBlock<Data>`, you can subscribe it directly to an `IObservable<T>`. And then you can configure whatever level of concurrency you want for that block.

But if you want to partition the data up and have concurrency across partitions, but serialized handling within partitions, and you want to use Rx to do that partitioning (which seems like a reasonable thing to want to do), then you'll need to think about how you'll structure the observers. You can't have a single observer, because the rules disallow concurrent invocations of OnNext on a single observer. You will therefore need one observer per partition. so you might need something broadly along these lines:

```cs
await dataItems
    .GroupBy(myKeySelector)
    .SelectMany(partition => partition
        .ObserveOn(new EventLoopScheduler())
        .Do(item =>
        {
            DoActualProcessing(item);
        }));
```

Here, we're using `GroupBy` to do the partitioning. It will emit an `IGroupedObservable` each time it sees a new key. 

We're then using `SelectMany`. The callback passed here will be invoked once for each partition. `SelectMany` will expect that callback to return an `IObservable<something>` and it will subscribe to whatever we return. In this case, I'm returning an observable based on the partition, but I've wrapped that in `ObserveOn` to indicate that I want each item that emerges from the partition to be handled through a particular scheduler. I'm creating an `EventLoopScheduler` here, which will spin up a new thread. It's important to construct a new one each time here, because this gives us the concurrency across partitions that I believe you're looking for.

When you use `ObserveOn`, it supplies an `IObserver<T>` to its source with an `OnNext` that doesn't block. It queues notifications up. So the `EventLoopScheduler` here essentially takes the place of what you were doing with the `Channel` objects. (Although if you particularly wanted to use `Channel`, you could write a `ChannelScheduler`. Notice that in this formulation it doesn't need to know what its input values are. We just create one scheduler per partition.)

So when the `GroupBy` operator delivers an item into the observer for a partition, that's going to go into the `ObserveOn` operator, which queues up a new work item with its scheduler and then returns immediately. The `EventLoopScheduler` for that partition will then process these work items one at a time for its partition, giving us the "one at a time within a partition" semantics. But since each partition will cause the `SelectMany` call back to run again, each partition gets its own `EventLoopScheduler`, giving us the "concurrent across partitions" behaviour that I think you're looking for.

So what then? Anything downstream of the `ObserveOn` is going to be running via the `EventLoopScheduler` here. So that `Select` that follows will invoke its callback on that scheduler's thread. Crucially, because we used `SelectMany`, it will subscribe to each observable we return separately. So each one gets its own `IObserver<T>` (supplied by the `SelectMany` operator). This is what frees us to process items concurrently: each partition has its own observer, so as long as it's one-at-a-time within any single partition, we're no longer breaking any rules.

I'm doing the actual work inside the `Do` callback. That's where our slow work is going. Because we set up the scheduling so that each partition gets its own `EventLoopScheduler`, it doesn't really matter if that's slow—the thread wasn't going to be doing anything else.

The `Do` operator is sort of like a tap: it lets you attach handlers to an observable like with `Subscribe` but it also returns an `IObservable<T>`. That's important here because `SelectMany` is going to expect us to return an `IObservable<something>`. So in practice, after each invocation of the `Do` callback, the observable returned by `Do` is going to emit its input. `SelectMany` is going to collect all of those, so it will flatten everything back out into a single `IObservable<Data>` that produces every single item. (As it happens it will produce them in the order they were processed, not the order they originally emerged.)

But all we're actually doing what that final, flattened, all-the-items observable is passing it to `await`. That subscribes to it and then waits for it to finish. (The `await` expression will also produce the very last item to emerge as its result, but we don't really care about that. We just want to know when it's done.)

So that's a more Rx-y way to do this. No need to write your own custom observable. No need to write your own custom scheduler (although if you really wanted to use `Channel` instead of the queue that `EventLoopScheduler` uses internally, you could). No need to break any of the rules.

However, one thing might throw a spanner in the works. What if instead of having a thread per partition, you wanted a single logical execution context, but you wanted to be able to use `async`/`await` so that you didn't need to create as many threads? Unfortunately, AsyncRx.NET isn't ready for prime time yet, so there's no good solution to that today.

So you'd end up having to do something slightly mess. Normally I wouldn't recommend nested calls to `Subscribe` like this next example does—where possible you want to let Rx manage the entire thing for you, like in my previous example. But to make this work for async workers, I think you'd have to do something a bit like this:

```cs
List<IDisposable> nestedSubs = new();
dataItems
    .GroupBy(myKeySelector)
    .Subscribe(
        partition =>
        {
            nestedSubs.Add(partition.Subscribe(new MyObserver());
        });
```

and then the `MyObserver` class mentioned here is a thing you'd need to write. That would implement `IObserver<T>`, and it would use `Channel` much like you're using internally to process the work. But instead of having that call onto some other `IObserver<T>.OnNext`, I'd have it just execute your work directly. You're outside of Rx at this point, and you need to be if you want to used `async`/`await` today.

By the way I think your use of `TaskCreationOptions.LongRunning` might be hurting you there, because the very first thing your task does is this:

```cs
await reader.WaitToReadAsync();
```

If there's nothing in the queue yet, this will immediately return, at which point the TPL will destroy the thread. When that `await` finally finishes, it will come in on an oridinary thread pool thread, not a long-running one. So you spun up a thread to execute one method and then immediately return.

**Comment by JanEggers** [comment]
thx ian for looking at my code and providing this detailed explanation.  I think you naild what I was trying to archive.

```c#
public static IObservable<TSource> ScheduleBy<TSource, TKey>(this IObservable<TSource> source, Func<TSource, TKey> keyselector)
    {
        return source
        .GroupBy(keyselector)
        .SelectMany(partition => partition
        .ObserveOn(new EventLoopScheduler()));
    }
```

this is doing exactly what i wanted.

**Comment by idg10** [comment]
Two things: 

I realised this morning that I missed out one very important thing with that example: the `EventLoopScheduler` instances will need to be `Dispose`d when everything has finished. Otherwise, the threads that these create will run forever, but will never be used again. The next code snippet fixes this with `Observable.Using`.

Second, your reduced version removes the call to `Do`, which was quite an important part. To illustrate why, I ran a little benchmark (which I'll show below) that produced these outputs:

```
ParallelPartitionedDo: 00:00:01.6606108

NonParallelPartitionedDo: 00:00:20.6939316
```

There was over a **12x difference in execution speed here** so this is quite an important difference! I have a 16 core machine, and I can see while running the `ParallelPartitionedDo` that all 16 cores get used, whereas the `NonParallelPartitionedDo` makes only 1 core busy. (One reason I'm not getting a 16x speedup is that the workload in this test is somewhat non-uniform, so not all 16 cores remain busy for the entire duration. But a 12x speedup is still pretty good.)

And the difference between these is extremely subtle:

```cs
static class RxEx
{
    // This is the one you want.
    public static IObservable<TSource> ParallelPartitionedDo<TSource, TKey>(
        this IObservable<TSource> source,
        Func<TSource, TKey> keyselector,
        Action<TSource> work)
    {
        return source
            .GroupBy(keyselector)
            .SelectMany(partition =>
                Observable.Using(
                    () => new EventLoopScheduler(),
                    scheduler => partition.ObserveOn(scheduler))
                .Do(work));
    }

    // This fails to parallelize calls to the work callback, so this runs many times
    // slower in scenarios where parallelism would have helped. (And it's basically
    // pointless: we partition, but then flatten the partitions back into a single
    // sequence before doing anything with the partitioning.)
    public static IObservable<TSource> NonParallelPartitionedDo<TSource, TKey>(
        this IObservable<TSource> source,
        Func<TSource, TKey> keyselector,
        Action<TSource> work)
    {
        return source
            .GroupBy(keyselector)
            .SelectMany(partition =>
                Observable.Using(
                    () => new EventLoopScheduler(),
                    scheduler => partition.ObserveOn(scheduler)))
            .Do(work);
    }
}
```

The difference is the placement of a single parenthesis. In the fast case, the `Do` is inside the callback passed to `SelectMany`, attached to the per-partition observable. In the slow case, it's applied to the `IObservable<T>` returned by `SelectMany`, which is the outer observable in which `SelectMany` has recombined all the outputs of the inner observables back into one observable.

The reason this makes such a huge difference is that to exploit parallelism, we need multiple `IObserver<T>`s. `SelectMany` creates an observer for each nested observable (for each partition in this case). And if we attach the `Do` to that, we get the parallelism we want.

But if we attach the `Do` further downstream—after `SelectMany` has flattened all those partitions back into one observable—we're back into a one-at-a-time mode because there is only one `Do` observer. `SelectMany` plays by Rx's rules: the combined `IObservable<T>` it returns will never have multiple calls to `OnNext` on its subscriber in progress simultaneously. It allows as much parallelism as the rules of Rx permit: so the individual observables created for each partition all get to run concurrently (as long as you supply a suitable scheduler, which we have here). But when they produce their outputs, `SelectMany` then uses thread synchronization techniques to ensure that if the inner observables produce items concurrently, the outputs come serially.

Here's the rest of the code for the benchmark. It renders the Mandelbrot set fractal as ASCII art (although by default it just prints out the benchmarks, because the ASCII art needed to be very large to create a sufficiently large quantity of work to be able to observe the effect of parallelism). It partitions the work by column—if you have a 16 core machine, the first partition will process all points with `X` coordinates (in the coordinate system of the output bitmap) of 0, 16, 32, 48, 64, etc., the next partition will handle all points with `X` coordinates of 1, 17, 33, 49, 65, and so on.

```cs
using System.Diagnostics;
using System.Numerics;
using System.Reactive.Concurrency;
using System.Reactive.Linq;

// Run this program with this argument:
//  /showmassiveoutput
// if you want a visualization of the work it's doing. Beware though
// that it requires a terminal window large enough to accommodate 256 columns
// of ASCII art!

bool SuppressMassiveAsciiArtOutput = args.Length == 0 || args[0] != "/showmassiveoutput";

// Non-square aspect ratio because in most terminals, characters are taller than they are wide.
const int Width = 256;
const int Height = 128;

// This frames the whole of the Mandelbrot set reasonably well.
const double centreX = -.75;
const double centreY = 0;
const double size = 1.25;

// This is an unnecessarily large number of iterations for rendering at this scale, but I've
// done this because we want to perform enough processing work to be able to see the benefits
// of parallelism.
const int MaxIterations = 100000;

// This just produces a grid of points, in the integer coordinate space of our output.
IObservable<(int X, int Y)> numbers =
    from x in Observable.Range(0, Width)
    from y in Observable.Range(0, Height)
    select (x, y);

int[,] results = new int[Width, Height];

// This is the fairly compute-intensive work we want to run for every single point.
void ProcessPoint((int X, int Y) p)
{
    // Map from the integer coordinate space of the output bitmap to the complex plane.
    double xr = ((double)p.X) / Width;
    double yr = ((double)p.Y) / Height;
    Complex c = new((xr - 0.5) * size * 2 + centreX, (yr - 0.5) * size * 2 + centreY);
    Complex x = 0;
    int count = 0;
    while (x.Magnitude < 10 && count < MaxIterations)
    {
        x = x * x + c;
        count += 1;
    }

    results[p.X, p.Y] = count;
}

Stopwatch sw = new();
sw.Start();
await numbers.ParallelPartitionedDo(
    p => p.X % Environment.ProcessorCount,
    ProcessPoint);
sw.Stop();

Console.WriteLine($"ParallelPartitionedDo: {sw.Elapsed}");
ShowAsAsciiArt();

sw.Start();
await numbers.NonParallelPartitionedDo(
    p => p.X % Environment.ProcessorCount,
    ProcessPoint);
sw.Stop();

Console.WriteLine();
Console.WriteLine($"NonParallelPartitionedDo: {sw.Elapsed}");
ShowAsAsciiArt();



void ShowAsAsciiArt()
{
    if (SuppressMassiveAsciiArtOutput) { return; }
    string scale = ".:-=+*#%@ ";

    StringBuilder sb = new(Width); 
    for (int y = 0; y < Height; y++)
    {
        sb.Clear();
        for (int x = 0; x < Width; x++)
        {
            double v = Math.Log(results[x, y]) / Math.Log(MaxIterations);
            sb.Append(scale[int.Min((int)(scale.Length * v), scale.Length - 1)]);
        }
        Console.WriteLine(sb.ToString());
    }
    Console.WriteLine();
}
```

**Comment by JanEggers** [comment]
thanks again for clarifiing
